Google Cloud: during Multipart - java.lang.IllegalArgumentException: Cannot create chunk from non-retainable ContentChunk

Jetty Version:- 12.0.4

Jetty Environment :- ee10

Java Version:- 17

While making a multipart request before getting it to the controller I'm getting the below error. Previously it was working with java 8, javax servlet api and jetty 9.


"Exception occur
org.springframework.web.multipart.MultipartException: Failed to parse multipart servlet request
	at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest.handleParseFailure(StandardMultipartHttpServletRequest.java:122)
	at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest.parseRequest(StandardMultipartHttpServletRequest.java:110)
	at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest.<init>(StandardMultipartHttpServletRequest.java:86)
	at org.springframework.web.multipart.support.StandardServletMultipartResolver.resolveMultipart(StandardServletMultipartResolver.java:112)
	at org.springframework.web.servlet.DispatcherServlet.checkMultipart(DispatcherServlet.java:1227)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1061)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:914)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:547)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:614)
	at org.eclipse.jetty.ee10.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1379)
	at org.eclipse.jetty.ee10.servlet.ServletHolder.handle(ServletHolder.java:736)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1614)
	at com.google.apphosting.utils.servlet.ee10.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:78)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.doFilter(FilterHolder.java:205)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1586)
	at com.google.apphosting.runtime.jetty.ee10.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:120)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.doFilter(FilterHolder.java:208)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1586)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$MappedServlet.handle(ServletHandler.java:1547)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.dispatch(ServletChannel.java:838)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:463)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:464)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:571)
	at org.eclipse.jetty.ee10.servlet.SessionHandler.handle(SessionHandler.java:703)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:761)
	at com.google.apphosting.runtime.jetty.ee10.AppEngineWebAppContext.handle(AppEngineWebAppContext.java:300)
	at org.eclipse.jetty.server.handler.HotSwapHandler.handle(HotSwapHandler.java:92)
	at org.eclipse.jetty.server.Handler$Wrapper.handle(Handler.java:716)
	at com.google.apphosting.runtime.jetty.CoreSizeLimitHandler.handle(CoreSizeLimitHandler.java:148)
	at org.eclipse.jetty.server.Server.handle(Server.java:179)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:597)
	at com.google.apphosting.runtime.jetty.delegate.internal.DelegateConnection.lambda$handle$0(DelegateConnection.java:145)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter$2.run(JettyServletEngineAdapter.java:118)
	at com.google.apphosting.runtime.jetty.delegate.internal.DelegateConnection.handle(DelegateConnection.java:141)
	at com.google.apphosting.runtime.jetty.delegate.DelegateConnector.service(DelegateConnector.java:49)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:223)
	at com.google.apphosting.runtime.RequestRunner.dispatchServletRequest(RequestRunner.java:262)
	at com.google.apphosting.runtime.RequestRunner.dispatchRequest(RequestRunner.java:227)
	at com.google.apphosting.runtime.RequestRunner.run(RequestRunner.java:193)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:273)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: jakarta.servlet.ServletException: org.eclipse.jetty.http.BadMessageException: 400: bad multipart
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getParts(ServletApiRequest.java:591)
	at jakarta.servlet.http.HttpServletRequestWrapper.getParts(HttpServletRequestWrapper.java:306)
	at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest.parseRequest(StandardMultipartHttpServletRequest.java:93)
	... 42 more
Caused by: org.eclipse.jetty.http.BadMessageException: 400: bad multipart
	... 45 more
Caused by: java.lang.IllegalArgumentException: Cannot create chunk from non-retainable ContentChunk@4c1ec711[l=true,b=HeapByteBuffer@7df9afb8[p=165,l=6634,c=6634,r=6469]={------Web...zip\r\n\r\n<<<PK\x03\x04\x14\x00\x00\x00\x08\x00\x16Z\x9dR\xEf?=L\x88\x15\x00\x00\x04\xA2\x00...aryVQ68AWpbxog4ELZy--\r\n>>>}]
	at org.eclipse.jetty.io.Content$Chunk.asChunk(Content.java:675)
	at org.eclipse.jetty.http.MultiPart$Parser.asSlice(MultiPart.java:1444)
	at org.eclipse.jetty.http.MultiPart$Parser.parseContent(MultiPart.java:1378)
	at org.eclipse.jetty.http.MultiPart$Parser.parse(MultiPart.java:1072)
	at org.eclipse.jetty.http.MultiPartFormData$Parser$1.parse(MultiPartFormData.java:229)
	at org.eclipse.jetty.http.MultiPartFormData$Parser$1.parse(MultiPartFormData.java:219)
	at org.eclipse.jetty.io.content.ContentSourceCompletableFuture.parse(ContentSourceCompletableFuture.java:104)
	at org.eclipse.jetty.http.MultiPartFormData$Parser.parse(MultiPartFormData.java:244)
	at org.eclipse.jetty.ee10.servlet.ServletMultiPartFormData.lambda$from$0(ServletMultiPartFormData.java:135)
	at org.eclipse.jetty.http.MultiPartFormData.from(MultiPartFormData.java:86)
	at org.eclipse.jetty.ee10.servlet.ServletMultiPartFormData.from(ServletMultiPartFormData.java:104)
	at org.eclipse.jetty.ee10.servlet.ServletMultiPartFormData.from(ServletMultiPartFormData.java:60)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getParts(ServletApiRequest.java:508)
	... 44 more
"

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MultiPartFormDataTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MultiPartFormDataTest.java
@@ -13,6 +13,7 @@
 
 package org.eclipse.jetty.http;
 
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.charset.Charset;
@@ -27,6 +28,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 
 import org.eclipse.jetty.io.Content;
 import org.eclipse.jetty.io.content.AsyncContent;
+import org.eclipse.jetty.io.content.InputStreamContentSource;
 import org.eclipse.jetty.toolchain.test.FS;
 import org.eclipse.jetty.toolchain.test.MavenTestingUtils;
 import org.eclipse.jetty.util.BufferUtil;
@@ -934,6 +936,48 @@ public class MultiPartFormDataTest
         assertInstanceOf(IllegalArgumentException.class, cause);
     }
 
+    @Test
+    public void testNonRetainableContent() throws Exception
+    {
+        String body = """
+            --AaB03x\r
+            content-disposition: form-data; name="field1"\r
+            \r
+            Joe Blow\r
+            --AaB03x\r
+            content-disposition: form-data; name="stuff"; filename="foo.txt"\r
+            Content-Type: text/plain\r
+            \r
+            aaaabbbbb\r
+            --AaB03x--\r
+            """;
+
+        Content.Source source = new InputStreamContentSource(new ByteArrayInputStream(body.getBytes(ISO_8859_1)))
+        {
+            @Override
+            public Content.Chunk read()
+            {
+                Content.Chunk chunk = super.read();
+                return new NonRetainableChunk(chunk);
+            }
+        };
+
+        MultiPartFormData.Parser formData = new MultiPartFormData.Parser("AaB03x");
+        formData.setFilesDirectory(_tmpDir);
+        try (MultiPartFormData.Parts parts = formData.parse(source).get(5, TimeUnit.SECONDS))
+        {
+            assertThat(parts.size(), is(2));
+            MultiPart.Part part1 = parts.getFirst("field1");
+            assertThat(part1, notNullValue());
+            Content.Source partContent = part1.getContentSource();
+            assertThat(Content.Source.asString(partContent), is("Joe Blow"));
+            MultiPart.Part part2 = parts.getFirst("stuff");
+            assertThat(part2, notNullValue());
+            partContent = part2.getContentSource();
+            assertThat(Content.Source.asString(partContent), is("aaaabbbbb"));
+        }
+    }
+
     private class TestContent extends AsyncContent
     {
         @Override
@@ -943,6 +987,45 @@ public class MultiPartFormDataTest
             if (chunk != null && chunk.canRetain())
                 _allocatedChunks.add(chunk);
             return chunk;
+        }
+    }
+
+    private static class NonRetainableChunk implements Content.Chunk
+    {
+        private final ByteBuffer _content;
+        private final boolean _isLast;
+        private final Throwable _failure;
+
+        public NonRetainableChunk(Content.Chunk chunk)
+        {
+            _content = BufferUtil.copy(chunk.getByteBuffer());
+            _isLast = chunk.isLast();
+            _failure = chunk.getFailure();
+            chunk.release();
+        }
+
+        @Override
+        public ByteBuffer getByteBuffer()
+        {
+            return _content;
+        }
+
+        @Override
+        public boolean isLast()
+        {
+            return _isLast;
+        }
+
+        @Override
+        public Throwable getFailure()
+        {
+            return _failure;
+        }
+
+        @Override
+        public void retain()
+        {
+            throw new UnsupportedOperationException();
         }
     }
 }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java
@@ -41,6 +41,8 @@ import org.eclipse.jetty.util.FutureCallback;
 import org.eclipse.jetty.util.FuturePromise;
 import org.eclipse.jetty.util.IO;
 import org.eclipse.jetty.util.Promise;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * <p>Namespace class that contains the definitions of a {@link Source content source},
@@ -48,6 +50,8 @@ import org.eclipse.jetty.util.Promise;
  */
 public class Content
 {
+    private static final Logger LOG = LoggerFactory.getLogger(Content.class);
+
     private Content()
     {
     }
@@ -666,15 +670,22 @@ public class Content
          * @param last whether the Chunk is the last one
          * @param retainable the Retainable this Chunk links to
          * @return a new Chunk
-         * @throws IllegalArgumentException if the {@code Retainable}
-         * {@link Retainable#canRetain() cannot be retained}
          */
         static Chunk asChunk(ByteBuffer byteBuffer, boolean last, Retainable retainable)
         {
-            if (!retainable.canRetain())
-                throw new IllegalArgumentException("Cannot create chunk from non-retainable " + retainable);
             if (byteBuffer.hasRemaining())
-                return new ByteBufferChunk.WithRetainable(byteBuffer, last, Objects.requireNonNull(retainable));
+            {
+                if (retainable.canRetain())
+                {
+                    return new ByteBufferChunk.WithRetainable(byteBuffer, last, Objects.requireNonNull(retainable));
+                }
+                else
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Copying buffer because could not retain");
+                    return new ByteBufferChunk.WithReferenceCount(BufferUtil.copy(byteBuffer), last);
+                }
+            }
             retainable.release();
             return last ? EOF : EMPTY;
         }

